### ISSUE TITLE
[2019 CVPR] AutoAugment: Learning Augmentation Strategies From Data

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#5]
Data augmentation 자동화 논문 중에서 가장 널리 알려진 논문

AutoAugment는 데이터로부터 직접 학습이 가능하며, transfer해도 여전히 random augmentation보다 좋다고 함

단점으로는 직접 학습 시 컴퓨팅 자원 및 시간 소모가 어마무시함
![image](https://user-images.githubusercontent.com/49643709/110773086-b4d1ee80-829f-11eb-9476-71fd57fe1591.png)

NAS와 매우 흡사한 방식으로 data augmentation을 학습함

전체적인 학습 pipeline은 다음과 같음

![Pipeline](https://user-images.githubusercontent.com/49643709/110741442-ffd80b80-8277-11eb-9ddd-4755a58aaebb.png)

1. Controller (RNN)이 data augmentation policy S를 sample
2. 해당 policy로 네트워크를 학습하여 validation accuracy R을 측정
3. Validation accuracy R을 지표로 강화학습을 통해 controller update

Data augmentation policy S는 5개의 sub-policy로 구성되었으며 sub-policy는 2개의 image operations sequence로 구성

Image operation은 probability, magnitude 총 2개의 hyperparameter를 가지고 있음

Image operation의 종류는 총 16가지임 : Shear X/Y, Translate X/Y, Rotate, AutoContrast, Invert, Equalize, Solarize, Posterize, Contrast, Color, Brightness, Sharpness, Cutout, Sample Pairing

![image](https://user-images.githubusercontent.com/49643709/110777143-2449dd00-82a4-11eb-91f2-6347c5f7c9ca.png)

Cutout 논문 : https://arxiv.org/pdf/1708.04552.pdf
Sample Pairing 논문 : https://arxiv.org/pdf/1801.02929.pdf

Magnitude는 대부분 실수이므로, 10개의 value로 discretize함 (uniform spacing)

Probability 또한 실수이므로, 11개의 value로 discretize함 (uniform spacing)

따라서 5 sub-policy의 search space는 (16 * 10 * 11)^(5 * 2)가 됨

이를 바탕으로 controller는 30 softmax prediction을 함 (5 sub-policy, 2 operation, 3 for operation type, magnitude, probability)

Validation accuracy를 지표로 강화학습 중 PPO (proximal policy optimization)를 통해 학습 진행

Controller가 softmax를 통해 data augmentation을 정해주면, 학습할 때 mini-batch에 있는 image마다 5개의 sub-policy 중 1개를 random 선택하여 학습 진행

Controller architecture는 100개의 hidden unit을 가진 one-layer LSTM이며, NASNet과 학습 방식이 매우 유사함

코드가 공개되어있지 않으며 네트워크 구조가 없어, 아마도 'operation type' -> 'magnitude' -> 'probability' 순서대로 10개의 operation을 예측하도록 구성했을 것으로 보임

![image](https://user-images.githubusercontent.com/49643709/110771906-62440280-829e-11eb-84c4-f06a70bbbbfa.png)
NASNet을 언급하였으나 N개의 replica를 생성하는 것이 아니므로 아마도 요런 느낌? (참고로 3부 Architecture of controller RNN and training hyperparamters에서 NASNet에 있는 내용을 그대로 복붙한 부분이 있는데 이 부분이 앞에 말한 30 softmax와 충돌)

Search가 끝나면 performance가 좋았던 5개의 policy를 합쳐서 1개로 concat한다 (총 25 sub-policy, 50 operation)

이를 이용하여 훈련을 진행

CIFAR-10/100 : CIFAR 10에서 4,000개 image random sample한 reduced-CIFAR-10에서 얻은 AutoAugment policy 이용

Baseline augmentation (standardize, random horizontal flip, zero-padding, random crop) -> AutoAugment policy -> Cutout (16 * 16) 순서로 augment하여 train

SVHN : 1,000개 image random sample한 dataset에서 얻은 AutoAugment policy 이용

Baseline augmentation (standardize) -> AutoAugment policy -> Cutout (20 * 20) 순서로 augment하여 train

ImageNet : 120개 class의 6,000 image sample한 dataset에서 얻은 AutoAugment policy 이용

Baseline augmentation (scale values, random horizontal flip) -> AutoAugment policy 순서로 augment하여 train

CIFAR, ImageNet은 상대적으로 color과 관련된 augmentation을 배우며 SVHN은 transform과 관련된 augmentation을 배움 

AutoAugment policy 결과는 다음과 같음

![image](https://user-images.githubusercontent.com/49643709/110777202-3461bc80-82a4-11eb-86c9-a5d5c03f8a53.png)
![image](https://user-images.githubusercontent.com/49643709/110777216-3a579d80-82a4-11eb-81bf-c124a7b41351.png)
![image](https://user-images.githubusercontent.com/49643709/110777234-3f1c5180-82a4-11eb-9e87-cd9cbb36ed4e.png)

실험 결과는 다음과 같음

![image](https://user-images.githubusercontent.com/49643709/110777311-55c2a880-82a4-11eb-90d5-76bea4a2779a.png)
![image](https://user-images.githubusercontent.com/49643709/110777333-5bb88980-82a4-11eb-917c-81eb12b9eb11.png)
![image](https://user-images.githubusercontent.com/49643709/110777348-6115d400-82a4-11eb-88ee-429873566896.png)
![image](https://user-images.githubusercontent.com/49643709/110777369-6541f180-82a4-11eb-92a9-e4b839e9afe2.png)
